src/CommonRuntime/Caching.fs

```diff
@@ -35,16 +36,19 @@ let createInMemoryCache expiration =
 open System.Collections.Concurrent
 
 /// Creates a cache that uses in-memory collection
-let createInMemoryCache expiration = 
+let createInMemoryCache (expiration:TimeSpan) = 
   let dict = new ConcurrentDictionary<_, _>()
   { new ICache<_> with
       member __.Set(key, value) = 
-        dict.[key] <- (value, DateTime.UtcNow)
+        dict.AddOrUpdate(key, (value, DateTime.UtcNow), (fun _ _ -> value, DateTime.UtcNow)) |> ignore
```

Comment:
What is the pont of using ConcurrentDictionary at the first place if you are not taking advantage of using GetOrAdd or AddOrUpdate methods which are guaranteeing the concistency by using the internal locks surrounding the lambda functions? I think that the abstraction level of the ICache interface is wrong.

The point of AddOrUpdate is to make it clear that on update-case the old value is being overridden by purpose. Of course it would be better to throw the ICache away and just use GetOrAdd.